prow/tide/tide.go

```diff
@@ -573,6 +581,90 @@ func filterPR(ghc githubClient, sp *subpool, pr *PullRequest) bool {
 	return false
 }
 
+// mergeMethodChecker provides a function to check if a PR can be merged with
+// the requested method and does not have a merge conflict.
+// It caches results and should be cleared periodically with clearCache()
+type mergeMethodChecker struct {
+	config config.Getter
+	ghc    githubClient
+	log    *logrus.Entry
+
+	sync.Mutex
+	cache map[config.OrgRepo]map[github.PullRequestMergeType]bool
+}
+
+func newMergeMethodChecker(cfg config.Getter, ghc githubClient, log *logrus.Entry) *mergeMethodChecker {
+	m := &mergeMethodChecker{
+		config: cfg,
+		ghc:    ghc,
+		log:    log,
+		cache:  map[config.OrgRepo]map[github.PullRequestMergeType]bool{},
+	}
+
+	go m.clearCache()
+	return m
+}
+
+func (m *mergeMethodChecker) clearCache() {
+	// Only do this once per token reset since it could be a bit expensive for
+	// Tide instances that handle hundreds of repos.
+	ticker := time.NewTicker(time.Hour)
+	for {
+		<-ticker.C
+		m.Lock()
+		m.cache = make(map[config.OrgRepo]map[github.PullRequestMergeType]bool)
+		m.Unlock()
+	}
+}
+
+func (m *mergeMethodChecker) repoMethods(orgRepo config.OrgRepo) (map[github.PullRequestMergeType]bool, error) {
+	m.Lock()
+	defer m.Unlock()
+
+	repoMethods, ok := m.cache[orgRepo]
+	if !ok {
+		fullRepo, err := m.ghc.GetRepo(orgRepo.Org, orgRepo.Repo)
+		if err != nil {
+			return nil, err
+		}
+		repoMethods = map[github.PullRequestMergeType]bool{
+			github.MergeMerge:  fullRepo.AllowMergeCommit,
+			github.MergeSquash: fullRepo.AllowSquashMerge,
+			github.MergeRebase: fullRepo.AllowRebaseMerge,
+		}
+		m.cache[orgRepo] = repoMethods
+	}
+	return repoMethods, nil
+}
+
+func (m *mergeMethodChecker) isAllowed(pr *PullRequest) error {
+	log := m.log.WithFields(pr.logFields())
+	if pr.Mergeable == githubql.MergeableStateConflicting {
+		return fmt.Errorf("PR has a merge conflict")
```

Comment:
If this also checks for conflicts, maybe rename the whole thing to `mergeChecker`? Having this check in the `mergeMethodChecker` is for me pretty unexpected.

Reply:
SGTM, I couldn't think of a good name. `mergeChecker` seems a little vague since it isn't clear what all it checks about mergeability, but it is certainly more sane and correct than what it is now. I'll make sure the go docs are clear.